### ISSUE TITLE
gulp-watch not reporting updated files

### ISSUE THREAD

**Issue by rhiadj** [floatdrop/gulp-watch#244]
Having done some digging the changes in the following two commits seem to have broken gulp-watch - or at least in the way I'm using it:
https://github.com/floatdrop/gulp-watch/commit/bc76bee
https://github.com/floatdrop/gulp-watch/commit/6cc9d40

**THE PROBLEM**
I have the following in a gulp build file (note the use of the "cwd" option):
```
gulp.task("watch", function () {
	watch('*.txt', { cwd: "app", verbose: true }, batch(function () {
		console.log("w00t!");
	}));
});
```

I'm running the following in the command line in the directory I'd expect gulp-watch to pick up the changes:
```
$ touch foo.txt
$ rm foo.txt
$ touch foo.txt
```

I'm seeing the following output when I run the ```watch``` task above:
```
[03:03:35] Using gulpfile /workspace/gulpfile.js
[03:03:35] Starting 'watch'...
[03:03:35] Finished 'watch' after 7.07 ms
```

... note there's **no** updates being reported by gulp-watch.

**POSSIBLE FIX**

I've made the following changes locally which seem to have fixed the issue (for me at least):
```
Index: index.js
===================================================================
--- index.js
+++ index.js
@@ -42,7 +42,12 @@
 		if (pathIsAbsolute(filepath)) {
 			return filepath;
 		}
-		return path.resolve(opts.cwd || process.cwd(), filepath);
+		else if (opts.cwd) {
+			return path.normalize(filepath);
+		}
+		else {
+			return path.resolve(process.cwd(), filepath);
+		}
 	}

 	function resolveGlob(glob) {
```

Now when I run the same above commands I'm seeing:
```
[03:05:53] Using gulpfile /workspace/gulpfile.js
[03:05:53] Starting 'watch'...
[03:05:53] Finished 'watch' after 6.27 ms
[03:06:01] foo.txt was added
w00t!
[03:06:05] foo.txt was unlinked
[03:06:17] foo.txt was added
```

... note there **are** now updates being reported by gulp-watch.

For now I've had to pin at 4.3.6 as that's the only version that works for me.

**Comment by UltCombo** [comment]
Hey, thanks for reporting and looking into the issue!

There are so many combinations of different options that it is very hard to get all of them right. Recently, I've been trying to normalize all inputted globs and watched paths to absolute paths, so they would work the same independently of the `cwd` option. This fixed some issues and didn't break any of the existing tests.

Due to all this complexity, every fix must have an accompanying unit test to prevent such regressions. Even then, it looks like we still have untested use cases.

Thanks for submitting a patch, but unfortunately I can't accept your suggested patch as it breaks 3 of our existing tests. Also, I've tried your exact example and I failed to reproduce the issue—I'm getting the added/unlinked events just fine with a clean install of v4.3.8.

Can you please try submitting a PR adding a failing unit test? You can base it off the [other tests](https://github.com/floatdrop/gulp-watch/blob/38938f698e98bf8ac82a2eeef351946aa70c5867/test/test-cwd.js#L20-L25). Also note that there's a nasty [Chokidar bug](https://github.com/paulmillr/chokidar/issues/449) in which it will never emit the ready event when one of your globs don't match anything. I've run into this bug a few times while writing tests.

**Comment by rhiadj** [comment]
Hi - thanks for the quick reply!

I've spent quite some time looking into this - and I almost can't believe what I've found! 😺 

Before I start I should mention I'm on OS X 10.10.1 in case that makes any difference.

I spent around a day debugging this in various ways:

1. I cloned ```gulp-watch``` into the same workspace directory as my other projects and ran the tests and they were pretty much all **failing**.

2. Debugged ```gulp-watch``` internally for around 3 hours and couldn't find a way to get any of the failing tests to pass, so gave up, deleted my cloned ```gulp-watch``` directory and took a break.

3. Came back and happened to reclone ```gulp-watch``` into ```/tmp/``` ran the tests and they all **passed** first time/every time!

4. Struck me it might be a permissions issue, so spent a while investigating permissions on every folder in the path to my workspace folder - all seemed fine.

5. Out of desperation I started copying the ```gulp-watch``` folder into every folder in the path, one by one, and running the tests - they **passed** in every folder up the path:
```
/Users/
/Users/Shared/
/Users/Shared/Stuff/
/Users/Shared/Stuff/Personal/
/Users/Shared/Stuff/Personal/Work/CompanyName
```
... until I got to the workspace folder:
```
/Users/Shared/Stuff/Personal/Work/CompanyName/Workspace/
```
... at this point the tests **failed** again.

6. Created a new folder alongside it:
```
/Users/Shared/Stuff/Personal/Work/CompanyName/Workspace2/
```
... and copied the entire  contents (including ```gulp-watch```) from ```Workspace``` to ```Workspace2```. At this point the tests all **passed**.

Doing an ```ls -la``` on the parent directory - all permissions/owner/group info is the same:
drwxr-xr-x   3 rhiad  wheel   136B 10 Jul 03:53 Workspace
drwxr-xr-x  10 rhiad  wheel   340B 10 Jul 03:53 Workspace2

7. Lastly I swapped the names around and the tests all still **passed**.

In a nutshell I have all the same files, in the same directory structure, with the same owners/groups/permissions as before - but now everything is magically working! I am utterly baffled!

All I can think is it was some kind of file system/hard drive/master file table glitch.

**Comment by rhiadj** [comment]
Closing this issue now - given everything is magically now working! 👍

**Comment by UltCombo** [comment]
Nice! Thanks for all the effort and throughout explanation. :smiley:

I honestly have no idea why that would happen, but maybe it could be some odd FSEvents/Chokidar bug? If you are still in the mood for testing, you could try setting `{ useFsEvents: false }` or `{ usePolling: true }` to check if it changes anything.

**Comment by rhiadj** [comment]
Thanks @UltCombo - I'm working to a deadline for Wednesday atm so won't get chance yet - I'll take a look once I've got that done and dusted 👍 ta
